### PR TITLE
Add connection-level security to frontend

### PR DESCRIPTION
**Pull request by JamesNK**
## Description

This PR adds connection-level security to the frontend. The frontend can now only be accessed via the frontend point.

This change is required because the OTLP HTTP endpoint supports being accessed via browser, and we don't want people to be able to browser the dashboard UI via the OTLP HTTP endpoint.

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [x] No - Will be discussed with OTLP HTTP changes
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5261)

### CODE REVIEW

**Comment by drewnoakes** [comment]
> This change is required because the OTLP HTTP endpoint supports being accessed via browser, and we don't want people to be able to browser the dashboard UI via the OTLP HTTP endpoint.

To confirm understanding, the issue before this PR was that a user can auth for OTLP HTTP then use that token to view the Dashboard?

**Comment by JamesNK** [comment]
I'm not sure. I didn't test what happens when someone accessed the browser dashboard when frontend auth was enabled. However, it was possible to view the dashboard via the OTLP HTTP port.

**Comment by JamesNK** [comment]
It would be useful if you double checked ODIC still works. The test passes, but I'm not sure if that completely verifies it is working. I believe there was a manual test you ran when that auth type was added.

File: src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        if (options.Frontend.AuthMode is FrontendAuthMode.OpenIdConnect)
+        {
+            return FrontendAuthenticationDefaults.AuthenticationSchemeOpenIdConnect;
+        }
+        else if (options.Frontend.AuthMode is FrontendAuthMode.BrowserToken)
+        {
+            return FrontendAuthenticationDefaults.AuthenticationSchemeBrowserToken;
+        }
+
+        return null;
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs:64]: Could be a switch expression.

```suggestion
        return dashboardOptions.CurrentValue.Frontend.AuthMode switch
        {
            FrontendAuthMode.OpenIdConnect => FrontendAuthenticationDefaults.AuthenticationSchemeOpenIdConnect,
            FrontendAuthMode.BrowserToken => FrontendAuthenticationDefaults.AuthenticationSchemeBrowserToken;
            _ => null;
        }
```

File: src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+{
+    protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        var result = await Context.AuthenticateAsync(ConnectionTypeAuthenticationDefaults.AuthenticationSchemeFrontend).ConfigureAwait(false);
+        if (result.Failure is not null)
+        {
+            return AuthenticateResult.Fail(
+                result.Failure,
+                new AuthenticationProperties(new Dictionary<string, string?>(),
+                new Dictionary<string, object?> { [AspirePolicyEvaluator.SuppressChallengeKey] = true }));
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs:28]: Indentation's bit off here. Named arguments would be helpful for me too.

```suggestion
            return AuthenticateResult.Fail(
                result.Failure,
                new AuthenticationProperties(
                    items: new Dictionary<string, string?>(),
                    parameters: new Dictionary<string, object?> { [AspirePolicyEvaluator.SuppressChallengeKey] = true }));
```

Also the `items` parameter is nullable, so can we avoid allocating an empty dictionary here?

File: src/Aspire.Dashboard/Authentication/Connection/ListenOptionsConnectionTypeExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+
+namespace Aspire.Dashboard.Authentication.Connection;
+
+internal static class ListenOptionsConnectionTypeExtensions
+{
+    public static void UseConnectionType(this ListenOptions listenOptions, ConnectionType[] connectionTypes)
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Authentication/Connection/ListenOptionsConnectionTypeExtensions.cs:10]: `UseConnectionTypes` (plural) feels more natural to me at call sites, given it takes multiple types.

**Review by drewnoakes — APPROVED** [review]
LGTM. Just some nits.

> I believe there was a manual test you ran when that auth type was added.

The manual test was for certificates, and Bala's team are running that test now. The OIDC tests here look good to me.

File: src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs (hunk 2 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs:28]: AuthenticationProperties allocates a dictionary if you pass in null anyway.

File: src/Aspire.Dashboard/Authentication/AspirePolicyEvaluator.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authorization.Policy;
+using Microsoft.AspNetCore.Authorization;
+using System.Security.Claims;
+
+namespace Aspire.Dashboard.Authentication;
+
+/// <summary>
+/// Mostly a copy of https://github.com/dotnet/aspnetcore/blob/e9647658dc46260cce21e980ec324565a4d39293/src/Security/Authorization/Policy/src/PolicyEvaluator.cs
+/// Type has been updated to support an auth scheme requesting that challenge is suppressed.
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Authentication/AspirePolicyEvaluator.cs:13]:
```suggestion
/// The logic has been updated to support an auth scheme requesting that challenge is suppressed.
```

File: src/Aspire.Dashboard/Authentication/Connection/ListenOptionsConnectionTypeExtensions.cs (hunk 3 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Authentication/Connection/ListenOptionsConnectionTypeExtensions.cs:10]:
```suggestion
    public static void UseConnectionTypes(this ListenOptions listenOptions, ConnectionType[] connectionTypes)
```

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 5]
```diff
@@ -508,6 +511,8 @@ static void AddEndpointConfiguration(Dictionary<string, string?> values, string
             {
                 configurationLoader.Endpoint(browserEndpointName, endpointConfiguration =>
                 {
+                    endpointConfiguration.ListenOptions.UseConnectionType([ConnectionType.Frontend]);
```
Comment (JamesNK) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:514]:
```suggestion
                    endpointConfiguration.ListenOptions.UseConnectionTypes([ConnectionType.Frontend]);
```

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 6]
```diff
@@ -528,10 +535,11 @@ static void AddEndpointConfiguration(Dictionary<string, string?> values, string
                             "The endpoint doesn't use TLS so browser access is only possible via a TLS terminating proxy.");
                     }
 
+                    connectionTypes.Add(ConnectionType.Frontend);
                     _frontendEndPointAccessor = _otlpServiceGrpcEndPointAccessor;
                 }
 
-                endpointConfiguration.ListenOptions.UseOtlpConnection();
+                endpointConfiguration.ListenOptions.UseConnectionType(connectionTypes.ToArray());
```
Comment (JamesNK) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:542]:
```suggestion
                endpointConfiguration.ListenOptions.UseConnectionTypes(connectionTypes.ToArray());
```

File: src/Aspire.Dashboard/DashboardWebApplication.cs [hunk 7]
```diff
@@ -557,10 +567,11 @@ static void AddEndpointConfiguration(Dictionary<string, string?> values, string
...
```
Comment (JamesNK) [inline src/Aspire.Dashboard/DashboardWebApplication.cs:574]:
```suggestion
                endpointConfiguration.ListenOptions.UseConnectionTypes(connectionTypes.ToArray());
```

File: src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    private string? GetRelevantAuthenticationScheme()
+    {
+        return dashboardOptions.CurrentValue.Frontend.AuthMode switch
+        {
+            FrontendAuthMode.OpenIdConnect => FrontendAuthenticationDefaults.AuthenticationSchemeOpenIdConnect,
+            FrontendAuthMode.BrowserToken => FrontendAuthenticationDefaults.AuthenticationSchemeBrowserToken;
+            _ => null;
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs:58]:
```suggestion
            FrontendAuthMode.BrowserToken => FrontendAuthenticationDefaults.AuthenticationSchemeBrowserToken,
            _ => null
```

File: src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    private string? GetRelevantAuthenticationScheme()
+    {
+        return dashboardOptions.CurrentValue.Frontend.AuthMode switch
+        {
+            FrontendAuthMode.OpenIdConnect => FrontendAuthenticationDefaults.AuthenticationSchemeOpenIdConnect,
+            FrontendAuthMode.BrowserToken => FrontendAuthenticationDefaults.AuthenticationSchemeBrowserToken,
+            _ => null
+        }
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Authentication/FrontendCompositeAuthenticationHandler.cs:59]:
```suggestion
        };
```

**Pull request merged**